examples/platforms/nrf52840/gpio.c

```diff
@@ -0,0 +1,236 @@
+/*
+ *  Copyright (c) 2017, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file:This file implements the OpenThread platform abstraction for GPIO
+ * and GPIOTE.
+ *
+ */
+
+#include "openthread-system.h"
+
+#include <string.h>
+
+#include <openthread/instance.h>
+
+#include "platform-nrf5.h"
+#include "hal/nrf_gpio.h"
+#include "hal/nrf_gpiote.h"
+#include "libraries/delay/nrf_delay.h"
+
+#if defined(__GNUC__)
+_Pragma("GCC diagnostic push")
+_Pragma("GCC diagnostic ignored \"-Wreturn-type\"")
+_Pragma("GCC diagnostic ignored \"-Wunused-parameter\"")
+_Pragma("GCC diagnostic ignored \"-Wunused-but-set-parameter\"")
+#endif
+#include "nrfx/drivers/include/nrfx_gpiote.h"
+#if defined(__GNUC__)
+_Pragma("GCC diagnostic pop")
+#endif
+
+/* Declaring callback functions for buttons 1 and 2 only. */
+otSysGpioIntCallback mHandler1;
+otSysGpioIntCallback mHandler2;
+
+/**
+ * @brief Function to receive interrupt and call back function
+ * set by the application for button 1.
+ *
+ */
+void in_pin1_handler(uint32_t pin, nrf_gpiote_polarity_t action)
+{
+    (void)pin;
+    (void)action;
+
+    /* Clear button interrupt */
+    otSysGpioIntClear(NRF_GPIOTE_EVENTS_PORT, pin);
+
+    /* Call the Button 1 handler registered by the application */
+    mHandler1(sInstance);
```

Comment:
This is being called from interrupt context. OpenThread APIs are designed to be called from a single thread of execution. In the bare-metal example, the preferred approach is to set a flag, and later have something like `otSysGpioProcess()` called from the main loop, which then checks the flag and calls the callback at that time.